src/Data/Array/Accelerate/Pattern/TH.hs

```diff
@@ -70,9 +70,16 @@ mkPattern nm = do
 mkDec :: Dec -> DecsQ
 mkDec dec =
   case dec of
-    DataD    _ nm tv _ cs _ -> mkDataD nm tv cs
-    NewtypeD _ nm tv _ c  _ -> mkNewtypeD nm tv c
+    DataD    _ nm tv _ cs _ -> mkDataD nm (fixTVList tv) cs
+    NewtypeD _ nm tv _ c  _ -> mkNewtypeD nm (fixTVList tv) c
     _                       -> fail "mkPatterns: expected the name of a newtype or datatype"
+  
+fixTV :: TyVarBndr a -> TyVarBndr ()
+fixTV (PlainTV n _ ) = PlainTV n ()
+fixTV (KindedTV n _ k) = KindedTV n () k
+
+fixTVList :: [TyVarBndr a] -> [TyVarBndr ()]
+fixTVList list = map fixTV list
```

Comment:
Quite unfortunate that the Functor instance of `TyVarBndr` is only there since template-haskell 2.17.0.0, which means [since GHC 9](https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/libraries/version-history)! Upping the lower bound to GHC 9 for this would indeed be a bit much :)

Reply:
I haven't tested, but I suspect that everything should also work if you replace the `()` by `a` in the type signatures of `mkDataD`, `mkNewtypeD`, `mkConP` and `mkConS` in that file